### ISSUE TITLE
SubprocVecEnv sometimes hangs if tensorflow is being used

### ISSUE THREAD

**Issue by christopherhesse** [openai/baselines#756]
This script seems to reproduce the hang pretty reliably on linux:

```
import multiprocessing
from baselines.common.vec_env.subproc_vec_env import SubprocVecEnv
import gym
import tensorflow as tf
import datetime


def loop():
    print(datetime.datetime.now().isoformat())
    # https://github.com/tensorflow/tensorflow/issues/20600
    graph = tf.Graph()  # not creating this graph avoids the hang
    sess = tf.Session(graph=graph)  
    with sess.graph.as_default():
        some_var = tf.Variable(0, dtype=tf.int64)
        SubprocVecEnv([lambda: gym.make('CartPole-v1') for i in range(1)])  # replacing with DummyVecEnv avoids the hang


def main():
    while True:
        loop()


if __name__ == '__main__':
    # this fixes the hang but is slower than the default fork method
    # https://github.com/tensorflow/tensorflow/issues/5448#issuecomment-314953888
    # multiprocessing.set_start_method('spawn')
    main()
```

It looks like tensorflow + fork() is not really supported, one possible fix is to use the `spawn` method instead.  There may be other possible fixes though.

**Comment by pzhokhov** [comment]
@christopherhesse which version of tensorflow? I am running it with 1.12.0, and it seems to work (well. I added venv.close - otherwise, the script was not hanging, but subprocesses were throwing EOFErrors). Also, spawn is apparently dramatically slower... We don't have to do it in a loop, so I suspect it is not too bad to use spawn, but I'd rather avoid it.

**Comment by christopherhesse** [comment]
Sorry it looks like this was merged to the internal repo before I saw this comment.  It happens pretty quickly when I run this with tensorflow 1.12.0 on linux.

It does seem to be slower, but spawning new vecenvs shouldn't happen that much anyway.  The fact that tensorflow is explicitly not fork-safe makes me want this to be the default despite the longer startup times though, since these rare sorts of hang bugs are kind of hard to debug.  It seems that john was encountering this error in his experiments, but it was fairly hard to reproduce.

If you're okay with the spawn fix, it looks like I also need to update shmemvecenv, since it also uses subprocesses.
